### ISSUE TITLE
Invalid password for streamers on older version of sam broacaster

### ISSUE THREAD

**Issue by Eyecu** [AzuraCast/AzuraCast#7562]
### Installation Method

Docker Installation

### AzuraCast Release Channel

Rolling Release Channel

### Current AzuraCast Version

#b56d6ed

### What happened?

Upgraded this afternoon to the latest rolling release, now all my DJ's on older version of sam broadcaster can't connect the encoder is reporting invalid password. .

EDIT: think this is the issue discussed here https://github.com/savonet/liquidsoap/issues/4210

### Relevant log output

```shell
2024/11/18 19:57:29 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:57:30 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 19:57:30 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:57:30 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 19:57:58 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:57:59 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 19:58:27 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:58:28 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 19:58:34 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:58:34 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 19:58:56 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:58:57 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 19:59:27 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
2024/11/18 19:59:27 [azuracast.api:3] API auth - Response (200): {"allow":false}
2024/11/18 20:00:31 [azuracast.api:3] API auth - Sending POST request to 'http://127.0.0.1:6010/api/internal/1/liquidsoap/auth' with body: { "user": "azriel", "password": "" }
```

**Comment by Eyecu** [comment]
Okay, weirdest thing, i was just playing around trying different things to see if I could get the older sam to login to the stream.
the setup is suppose to be user,password in the encoder password field but that was giving the password error, just by chance I added a test like user,user,password and it logs in so like wes,wes,password and that worked it logs in and streams.

**Comment by Aealyn** [comment]
It's good to know. But very few people will understand that. Therefore, in my opinion, this should be changed back to the old variant user, password.

**Comment by tatokis** [comment]
While I don't use AzuraCast, I had a quick look and the code seems to describe exactly the behaviour you're observing.

https://github.com/AzuraCast/AzuraCast/blob/c8149993213dcf409a9a1927fbb8241d9fd38f6e/util/docker/stations/liquidsoap/azuracast.liq#L445-L451

I'm not sure how that worked in the past, but without knowing why it was done that way, I think that `2` at the end of line 448 should probably be a `1`.

That code above also suffers from the same issue as https://github.com/savonet/liquidsoap/issues/4210, i.e. if you set the password field to "username,pa:ss,word" it will only keep "pa" as the password.

A more correct (untested) implementation could be something like
```
sep_re = r/([,:])/
auth_info =
    if (login.user == "source" or login.user == "") and (sep_re.test(login.password)) then
        sep = sep_re.exec(login.password)
        let (u, p) = string.split.first(separator=sep[0], login.password)
        {user = u, password = p}
    else
        {user = login.user, password = login.password}
    end
```

**Comment by Aealyn** [comment]
Even after the latest updates, connecting via SAM 3 or SAM 4.2.2 is not possible without problems.

Only through the complicated username, username, password. Please please try to fix this error, otherwise I won't be able to make updates, which would be a shame. Because first explaining to everyone how they have to connect, hardly anyone will understand that.

**Comment by BusterNeece** [comment]
@Eyecu @Aealyn So it turns out @tatokis was onto something here...I looked at our DJ auth code and also wondered why we were referencing indices 0 and 2, when intuitively I'd reference 0 and 1 on a string split...

I talked it over with the Liquidsoap folks, and it turns out that in previous versions, there was something of a "bug" in how this worked with regex-based string separators, where it would return the actual string separator as its own item, so the return was `("username", ":", "password")`, not the more intuitive `("username", "password")`. This was changed to be more consistent in 2.3.x, and the change is honestly more intuitive, so I updated our code to reflect it.

The latest Rolling Release that's deploying now shouldn't require the weird `username,username,password` thing anymore. Thanks folks for your patience.
